pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -422,6 +482,45 @@ func (t *ResolvedPipelineTask) skipBecauseResultReferencesAreMissing(facts *Pipe
 	return false
 }
 
+// skipBecausePipelineRunPipelineTimeoutReached returns true if the task shouldn't be launched because the elapsed time since
+// the PipelineRun started is greater than the PipelineRun's pipeline timeout
+func (t *ResolvedPipelineTask) skipBecausePipelineRunPipelineTimeoutReached(facts *PipelineRunFacts) bool {
+	if t.checkParentsDone(facts) {
+		if facts.TimeoutsState.PipelineTimeout != nil && *facts.TimeoutsState.PipelineTimeout != config.NoTimeoutDuration && facts.TimeoutsState.StartTime != nil {
+			// If the elapsed time since the PipelineRun's start time is greater than the the PipelineRun's Pipeline timeout, skip.
+			return facts.TimeoutsState.Clock.Since(*facts.TimeoutsState.StartTime) > *facts.TimeoutsState.PipelineTimeout
+		}
+	}
+
+	return false
+}
+
+// skipBecausePipelineRunTasksTimeoutReached returns true if the task shouldn't be launched because the elapsed time since
+// the PipelineRun started is greater than the PipelineRun's tasks timeout
+func (t *ResolvedPipelineTask) skipBecausePipelineRunTasksTimeoutReached(facts *PipelineRunFacts) bool {
+	if t.checkParentsDone(facts) && !t.IsFinalTask(facts) {
+		if facts.TimeoutsState.TasksTimeout != nil && *facts.TimeoutsState.TasksTimeout != config.NoTimeoutDuration && facts.TimeoutsState.StartTime != nil {
+			// If the elapsed time since the PipelineRun's start time is greater than the the PipelineRun's Tasks timeout, skip.
+			return facts.TimeoutsState.Clock.Since(*facts.TimeoutsState.StartTime) > *facts.TimeoutsState.TasksTimeout
+		}
+	}
+
+	return false
+}
+
+// skipBecausePipelineRunFinallyTimeoutReached returns true if the task shouldn't be launched because the elapsed time since
+// finally tasks started being executed is greater than the PipelineRun's finally timeout
```

Comment:
> `task shouldn't be launched`

which task shouldn't be launched? all finally tasks are suppose to start at the same time, will there any tasks left to meet this criteria? If the finally tasks are started and the finally time out hits, are they getting cancelled or skipped?

Reply:
In practice, this is probably only going to hit in some really bizarre edge case, but I feel like it’s better to be safe and keep handling past-timeout finally tasks the same way we do dag tasks, since that’s what we’re doing with the 1s timeout approach currently. I’d be open to just not bothering with finally tasks and letting the PR handle timing out all cases of executing finally tasks when we’re past the finally timeout.